File: pdui_flutter/lib/widgets/pdui_app_bar.dart
import 'package:flutter/material.dart';

import '../proto_out/pdui-proto-out.pb.dart';
import 'pdui_widget.dart';

class PduiAppBar extends PduiWidget {
  PBExpression pbAppBar;

  PduiAppBar(this.pbAppBar);

  @override
  AppBar resolve() {
    late Color backgroundColor = Color(int.parse("00000000", radix: 16));

    if (pbAppBar.appBar.backgroundColor != "") {
      String argbColor = pbAppBar.appBar.backgroundColor.replaceAll("#", "");
      backgroundColor = Color(int.parse(argbColor, radix: 16));
    }

    return AppBar(
      key: pbAppBar.key != "" ? Key(pbAppBar.key) : null,
      title: Text(pbAppBar.appBar.title),
      backgroundColor: backgroundColor,
    );
  }
}

File: pdui-express/lib/widgets/PDUIAppBar.ts
import { PDUIWidget } from "../models/PDUIWidget";
import type { PBAppBar, PBExpression } from "../proto-out/pdui-proto-out.ts";

export class PDUIAppBar extends PDUIWidget {
    title: string;
    //TODO: Docs
    //Format: FF000000 (alpha, r, g, b)
    backgroundColor: string;

    constructor({
        key,
        title,
        backgroundColor,
    }: { key?: string; title: string; backgroundColor: string }) {
        //TODO: Validate the backgroundColor format
        super({
            key,
            identity: "appBar",
            payload: {
                oneofKind: "appBar",
                appBar: {
                    title: title,
                    backgroundColor: backgroundColor,
                },
            },
        });

        this.title = title;
        this.backgroundColor = backgroundColor;
    }

    override toPB(): PBAppBar {
        return { title: this.title, backgroundColor: this.backgroundColor };
    }

    override toExpression(): PBExpression {
        return {
            key: this.key,
            identity: this.identity,
            payload: {
                oneofKind: "appBar",
                appBar: {
                    title: this.title,
                    backgroundColor: this.backgroundColor,
                },
            },
        };
    }
}
